### PR TITLE
Prevent removal of admin and default members from workspace chats

### PR DESCRIPTION
**Pull request by brunovjk**
### Details

This PR addresses an [issue](https://github.com/Expensify/App/issues/35391#issuecomment-2140731026) where **workspace admins** and **default-added-users** were inadvertently selectable for removal from **workspace chats**. To fix this, the PR implements controls based on user roles, as documented in [workspace rooms](https://github.com/Expensify/App/blob/main/README.md#workspace-rooms) and [workspace chats](https://github.com/Expensify/App/blob/main/README.md#workspace-chats).

### Fixed Issues

$ https://github.com/Expensify/App/issues/35391
PROPOSAL: N/A

### Pre-condition:

**1.** We need 4 accounts for this test:
- Workspace Creator
- Workspace Admin
- Workspace Member
- No-workspace member (invited)

**2.** Login in the app as _"Workspace Creator"_
**3.** Create a workspace
**4.** Go to the automatically created _"Workspace Chat"_
**5.** Go to the _"Report Details"_
**6.** Go to _"Members"_ page > Click _"Invite"_
**7.** Invite all the other 3 accounts (`admin`, `member`, and `invited`) to the _"Workspace Chat"_
**8.** Still on the _"Members"_ page, change the `admin` role to `Administrator`
**8.** Click on the _"Global Fab"_ > _"Start Chat"_ > Select _"#Room"_ tab
**9.** Create a _"workspace-room"_
**10.** Invite all the other 3 accounts (`admin`, `member`, and `invited`) to the _"workspace-room"_

### Tests:

**1. Logged as the "Workspace Creator"**:

   **i.** Go to the "Workspace Chat" > Report Details page
   **ii.** Verify that the "Leave" button doesn't show, as you are a "workspace admin".
   **iii.** Go to "Members" page
   **iv.** Verify that you cannot select/remove admin and default added members
   **v.** Verify that you can "Invite" and "Remove" (admin cannot be removed) members
   **vi.** Go to the "workspace-room" > Report Details page
   **vii**. Go to "Members" page
   **viii.** Verify that you can "Invite" and "Remove" (you cannot remove yourself here) members

**2. Logged as the "Workspace Admin"**:

   **i.** Go to the "Workspace Chat" > Report Details page
   **ii.** Verify that the "Leave" button doesn't show, as you are a "workspace admin".
   **iii.** Go to "Members" page
   **iv.** Verify that you cannot select/remove admin and default added members
   **v.** Verify that you can "Invite" and "Remove" (admin cannot be removed) members
   **vi.** Go to the "workspace-room" > Report Details page
   **vii.** Go to "Members" page
   **viii.** Verify that you can "Invite" and "Remove" (you cannot remove yourself here) members

**3. Logged as the "Workspace Member"**:

   **i.** Go to the "Workspace Chat" > Report Details page
   **ii.** Verify that the "Invite" button does not show.
   **iii.** Go to "Members" page
   **iv.** Verify that you cannot "Invite" and "Remove" members.
   **vii.** Logged as the "Workspace Member" > Go to the "workspace-room" > Report Details page
   **viii.** Verify that the "Invite" button does not show.
   **x.** Go to "Members" page
   **xi.** Verify that you can "Invite" and "Remove" (you cannot remove yourself here) members.

**4. Logged as the "Workspace Chat Invited**":

   **i.** Go to the "Workspace Chat" > Report Details page
   **ii.** Verify that the "Invite" button does not show.
   **iii.** Go to "Members" page
   **iv.** Verify that you cannot "Invite" and "Remove" members.
   **vii.** Logged as the "Workspace Room Invited" > Go to the "workspace-room" > Report Details page
   **viii.** Verify that the "Invite" button does show and you can invite members to the chat.

- [x] Verify that no errors appear in the JS console

### Offline tests

N/A

### QA Steps

Same as Tests

### PR Author Checklist

- [x] I linked the correct issue in the `### Fixed Issues` section above
- [x] I wrote clear testing steps that cover the changes made in this PR
    - [x] I added steps for local testing in the `Tests` section
    - [x] I added steps for the expected offline behavior in the `Offline steps` section
    - [x] I added steps for Staging and/or Production testing in the `QA steps` section
    - [x] I added steps to cover failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
    - [x] I tested this PR with a [High Traffic account](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#high-traffic-accounts) against the staging or production API to ensure there are no regressions (e.g. long loading states that impact usability).
- [x] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [x] I ran the tests on **all platforms** & verified they passed on:
    - [x] Android: Native
    - [x] Android: mWeb Chrome
    - [x] iOS: Native
    - [x] iOS: mWeb Safari
    - [x] MacOS: Chrome / Safari
    - [x] MacOS: Desktop
- [x] I verified there are no console errors (if there's a console error not related to the PR, report it or open an issue for it to be fixed)
- [x] I followed proper code patterns (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`)
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
      - [x] If any non-english text was added/modified, I verified the translation was requested/reviewed in #expensify-open-source and it was approved by an internal Expensify engineer. Link to Slack message:
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I followed the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I tested other components that can be impacted by my changes (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` are working as expected)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] I verified that if a function's arguments changed that all usages have also been updated correctly
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/styles/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG)`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>

https://github.com/Expensify/App/assets/95647348/61582245-2417-436b-9498-55cfdd3a1a81

</details>

<details>
<summary>Android: mWeb Chrome</summary>

https://github.com/Expensify/App/assets/95647348/0dd0572e-f3f7-40b8-8398-dffa387c3d83

</details>

<details>
<summary>iOS: Native</summary>

https://github.com/Expensify/App/assets/95647348/34d7a544-1859-43c3-ac24-c100d0713285

</details>

<details>
<summary>iOS: mWeb Safari</summary>

https://github.com/Expensify/App/assets/95647348/b3929bfc-e7f6-4c8d-9a80-9ece68d077bd

</details>

<details>
<summary>MacOS: Chrome / Safari</summary>

https://github.com/Expensify/App/assets/95647348/7a667544-00c3-430b-b38b-f62aa7582485

</details>

<details>
<summary>MacOS: Desktop</summary>

https://github.com/Expensify/App/assets/95647348/9dfe64a0-c5cb-4054-a953-812844f510df

</details>

### CODE REVIEW

**Comment by mananjadhav** [comment]
@brunovjk can you fix the lint issues? I can then review the code.

**Comment by brunovjk** [comment]
> @brunovjk can you fix the lint issues? I can then review the code.

This lint issue wasn't caused by us. It even has a reported [issue](https://github.com/Expensify/App/issues/43326) for that. I fixed it anyway, but I can revert if needed. Thank you!

**Comment by brunovjk** [comment]
All good here @mananjadhav :D

**Comment by mananjadhav** [comment]
Thanks @brunovjk for fixing this. I am currently testing it now.

**Comment by mananjadhav** [comment]
## Reviewer Checklist

- [x] I have verified the author checklist is complete (all boxes are checked off).
- [x] I verified the correct issue is linked in the `### Fixed Issues` section above
- [x] I verified testing steps are clear and they cover the changes made in this PR
    - [x] I verified the steps for local testing are in the `Tests` section
    - [x] I verified the steps for Staging and/or Production testing are in the `QA steps` section
    - [x] I verified the steps cover any possible failure scenarios (i.e. verify an input displays the correct error message if the entered data is not correct)
    - [x] I turned off my network connection and tested it while offline to ensure it matches the expected behavior (i.e. verify the default avatar icon is displayed if app is offline)
- [x] I checked that screenshots or videos are included for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I included screenshots or videos for tests on [all platforms](https://github.com/Expensify/App/blob/main/contributingGuides/CONTRIBUTING.md#make-sure-you-can-test-on-all-platforms)
- [ ] I verified tests pass on **all platforms** & I tested again on:
    - [ ] Android: Native
    - [ ] Android: mWeb Chrome
    - [ ] iOS: Native
    - [ ] iOS: mWeb Safari
    - [ ] MacOS: Chrome / Safari
    - [ ] MacOS: Desktop
- [x] If there are any errors in the console that are unrelated to this PR, I either fixed them (preferred) or linked to where I reported them in Slack
- [x] I verified proper code patterns were followed (see [Reviewing the code](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md#reviewing-the-code))
    - [x] I verified that any callback methods that were added or modified are named for what the method does and never what callback they handle (i.e. `toggleReport` and not `onIconClick`).
    - [x] I verified that the left part of a conditional rendering a React component is a boolean and NOT a string, e.g. `myBool && <MyComponent />`.
    - [x] I verified that comments were added to code that is not self explanatory
    - [x] I verified that any new or modified comments were clear, correct English, and explained "why" the code was doing something instead of only explaining "what" the code was doing.
    - [x] I verified any copy / text shown in the product is localized by adding it to `src/languages/*` files and using the [translation method](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60)
    - [x] I verified all numbers, amounts, dates and phone numbers shown in the product are using the [localization methods](https://github.com/Expensify/App/blob/4bd99402cebdf4d7394e0d1f260879ea238197eb/src/components/withLocalize.js#L60-L68)
    - [x] I verified any copy / text that was added to the app is grammatically correct in English. It adheres to proper capitalization guidelines (note: only the first word of header/labels should be capitalized), and is either coming verbatim from figma or has been approved by marketing (in order to get marketing approval, ask the Bug Zero team member to add the Waiting for copy label to the issue)
    - [x] I verified proper file naming conventions were followed for any new files or renamed files. All non-platform specific files are named after what they export and are not named "index.js". All platform-specific files are named for the platform the code supports as outlined in the README.
    - [x] I verified the JSDocs style guidelines (in [`STYLE.md`](https://github.com/Expensify/App/blob/main/contributingGuides/STYLE.md#jsdocs)) were followed
- [x] If a new code pattern is added I verified it was agreed to be used by multiple Expensify engineers
- [x] I verified that this PR follows the guidelines as stated in the [Review Guidelines](https://github.com/Expensify/App/blob/main/contributingGuides/PR_REVIEW_GUIDELINES.md)
- [x] I verified other components that can be impacted by these changes have been tested, and I retested again (i.e. if the PR modifies a shared library or component like `Avatar`, I verified the components using `Avatar` have been tested & I retested again)
- [x] I verified all code is DRY (the PR doesn't include any logic written more than once, with the exception of tests)
- [x] I verified any variables that can be defined as constants (ie. in CONST.js or at the top of the file that uses the constant) are defined as such
- [x] If a new component is created I verified that:
    - [x] A similar component doesn't exist in the codebase
    - [x] All props are defined accurately and each prop has a `/** comment above it */`
    - [x] The file is named correctly
    - [x] The component has a clear name that is non-ambiguous and the purpose of the component can be inferred from the name alone
    - [x] The only data being stored in the state is data necessary for rendering and nothing else
    - [x] For Class Components, any internal methods passed to components event handlers are bound to `this` properly so there are no scoping issues (i.e. for `onClick={this.submit}` the method `this.submit` should be bound to `this` in the constructor)
    - [x] Any internal methods bound to `this` are necessary to be bound (i.e. avoid `this.submit = this.submit.bind(this);` if `this.submit` is never passed to a component event handler like `onClick`)
    - [x] All JSX used for rendering exists in the render method
    - [x] The component has the minimum amount of code necessary for its purpose, and it is broken down into smaller components in order to separate concerns and functions
- [x] If any new file was added I verified that:
    - [x] The file has a description of what it does and/or why is needed at the top of the file if the code is not self explanatory
- [x] If a new CSS style is added I verified that:
    - [x] A similar style doesn't already exist
    - [x] The style can't be created with an existing [StyleUtils](https://github.com/Expensify/App/blob/main/src/utils/index.ts) function (i.e. `StyleUtils.getBackgroundAndBorderStyle(theme.componentBG`)
- [x] If the PR modifies code that runs when editing or sending messages, I tested and verified there is no unexpected behavior for all supported markdown - URLs, single line code, code blocks, quotes, headings, bold, strikethrough, and italic.
- [x] If the PR modifies a generic component, I tested and verified that those changes do not break usages of that component in the rest of the App (i.e. if a shared library or component like `Avatar` is modified, I verified that `Avatar` is working as expected in all cases)
- [x] If the PR modifies a component related to any of the existing Storybook stories, I tested and verified all stories for that component are still working as expected.
- [x] If the PR modifies a component or page that can be accessed by a direct deeplink, I verified that the code functions as expected when the deeplink is used - from a logged in and logged out account.
- [x] If the PR modifies the UI (e.g. new buttons, new UI components, changing the padding/spacing/sizing, moving components, etc) or modifies the form input styles:
    - [x] I verified that all the inputs inside a form are aligned with each other.
    - [x] I added `Design` label and/or tagged `@Expensify/design` so the design team can review the changes.
- [x] If a new page is added, I verified it's using the `ScrollView` component to make it scrollable when more elements are added to the page.
- [x] If the `main` branch was merged into this PR after a review, I tested again and verified the outcome was still expected according to the `Test` steps.
- [x] I have checked off every checkbox in the PR reviewer checklist, including those that don't apply to this PR.

### Screenshots/Videos
<details>
<summary>Android: Native</summary>



</details>

<details>
<summary>Android: mWeb Chrome</summary>



</details>

<details>
<summary>iOS: Native</summary>



</details>

<details>
<summary>iOS: mWeb Safari</summary>



</details>

<details>
<summary>MacOS: Chrome / Safari</summary>



</details>

<details>
<summary>MacOS: Desktop</summary>



</details>

**Review by mananjadhav — APPROVED** [review]

File: src/pages/ReportDetailsPage.tsx [hunk 1]
```diff
@@ -160,7 +160,7 @@ function ReportDetailsPage({policies, report, session, personalDetails}: ReportD
                 isAnonymousAction: false,
                 shouldShowRightIcon: true,
                 action: () => {
-                    if (isUserCreatedPolicyRoom || isChatThread || isPolicyExpenseChat) {
+                    if (isUserCreatedPolicyRoom || isChatThread || (isPolicyExpenseChat && isPolicyAdmin)) {
```
Comment (jasperhuangg) [inline src/pages/ReportDetailsPage.tsx:163]: Why are we preventing people other than policy admins from accessing the members page?

Comment (jasperhuangg) [reply src/pages/ReportDetailsPage.tsx:163]: I thought this change only involved preventing the removal of default members from workspace chats.

**Review by jasperhuangg — CHANGES_REQUESTED** [review]

File: src/pages/ReportDetailsPage.tsx (hunk 1 above)
Comment (brunovjk) [reply src/pages/ReportDetailsPage.tsx:163]: > Why are we preventing people other than policy admins from accessing the members page?

_"Invited members(invited by members) are not able to invite or remove from the workspace chat"_ 
_"Default members and admins are able to remove invited members"_
https://github.com/Expensify/App?tab=readme-ov-file#workspace-chats

I believe only _Admin_ can access Members page, others access Participants page.

Comment (brunovjk) [reply src/pages/ReportDetailsPage.tsx:163]: > I thought this change only involved preventing the removal of default members from workspace chats.

Sure. _"Admins are not able to leave/be removed from the workspace chat"_.
https://github.com/Expensify/App?tab=readme-ov-file#workspace-chats

I noticed that automatically added users, example _guides_ accounts, are marked as _POLICY.ROLE.ADMIN_.
Therefore, I believe that this check is necessary and enough. Thoughts?

**Comment by brunovjk** [comment]
Resolve conflicts.
Tested all the flows described in README [here](https://github.com/Expensify/App?tab=readme-ov-file#workspace-chats), and everything looks good to me.

**Note**: We still have a type error that is being resolved [here](https://github.com/Expensify/App/issues/43710), will "merge main" when fixed.

File: src/pages/ReportDetailsPage.tsx (hunk 1 above)
Comment (jasperhuangg) [reply src/pages/ReportDetailsPage.tsx:163]: Ah okay got it, thanks for clarifying, I'm fine with this change then 🙇

**Comment by brunovjk** [comment]
All yours @jasperhuangg  :D

**Review by jasperhuangg — APPROVED** [review]

**Comment by kbecciv** [comment]
This PR is failing because of issue https://github.com/Expensify/App/issues/35391

The issue is reproducible in: Web

https://github.com/Expensify/App/assets/93399543/de0a01f7-755b-4a84-b3bf-833cfdf0b555

**Comment by brunovjk** [comment]
> This PR is failing because of issue #35391
> 
> The issue is reproducible in: Web
> 
>  Screen.Recording.2024-06-16.at.6.50.53.in.the.evening.mp4

I'm sorry, but the test steps were wrong. I fixed it now:

3. Logged as the "Workspace Member":
...
... Go to the "workspace-room"
- xi. Verify that you can**not** "Invite" and "Remove" (you cannot remove yourself here) members.
+ xi. Verify that you can "Invite" and "Remove" (you cannot remove yourself here) members.

**Comment by kbecciv** [comment]
PR is Pass according to the new step https://github.com/Expensify/App/pull/43065#issuecomment-2173968111

https://github.com/Expensify/App/assets/93399543/eb50207d-b177-4bf5-aab0-36dfa8ca6839

File: src/pages/RoomMembersPage.tsx [hunk 2]
```diff
@@ -182,12 +184,14 @@ function RoomMembersPage({report, session, policies}: RoomMembersPageProps) {
                 return;
             }
             const pendingChatMember = report?.pendingChatMembers?.findLast((member) => member.accountID === accountID.toString());
+            const isAdmin = !!(policy && policy.employeeList && details.login && policy.employeeList[details.login]?.role === CONST.POLICY.ROLE.ADMIN);
+            const isDisabled = (isPolicyExpenseChat && isAdmin) || accountID === session?.accountID || pendingChatMember?.pendingAction === CONST.RED_BRICK_ROAD_PENDING_ACTION.DELETE;
```
Comment (jayeshmangwani) [inline src/pages/RoomMembersPage.tsx:188]: Regarding this issue https://github.com/Expensify/App/issues/42928, when determining if the option should be disabled, we should have also checked the workspace chat's ownerAccountID and disable it when account id matches the ownerAccountID. 

We [fixed](https://github.com/Expensify/App/pull/44595) this by adding a check `details.accountID === report.ownerAccountID;`

**Pull request merged**